controllers/storagecluster/storageclient.go

```diff
@@ -0,0 +1,160 @@
+package storagecluster
+
+import (
+	"crypto/tls"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"time"
+
+	ocsclientv1a1 "github.com/red-hat-storage/ocs-client-operator/api/v1alpha1"
+	ocsv1 "github.com/red-hat-storage/ocs-operator/api/v4/v1"
+	ocsinit "github.com/red-hat-storage/ocs-operator/v4/controllers/ocsinitialization"
+	extv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	cutil "sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+const (
+	storageClientCRDName   = "storageclients.ocs.openshift.io"
+	storageClientName      = "local-storageclient"
+	onboardingTokenURLPath = "onboarding-tokens"
+	saTokenPath            = "/var/run/secrets/kubernetes.io/serviceaccount/token"
+)
+
+type storageClient struct{}
+
+var _ resourceManager = &storageClient{}
+
+func (s *storageClient) ensureCreated(r *StorageClusterReconciler, storagecluster *ocsv1.StorageCluster) (reconcile.Result, error) {
+	if !storagecluster.Spec.AllowRemoteStorageConsumers {
+		r.Log.Info("Spec.AllowRemoteStorageConsumers is disabled")
+		return reconcile.Result{}, nil
+	}
+
+	if res, err := s.createStorageClient(r, storagecluster); err != nil {
+		return reconcile.Result{}, err
+	} else if !res.IsZero() {
+		return res, nil
+	}
+
+	return reconcile.Result{}, nil
+}
+
+func (s *storageClient) ensureDeleted(r *StorageClusterReconciler, storagecluster *ocsv1.StorageCluster) (reconcile.Result, error) {
+	if !storagecluster.Spec.AllowRemoteStorageConsumers {
+		r.Log.Info("Spec.AllowRemoteStorageConsumers is disabled")
+		return reconcile.Result{}, nil
+	}
+
+	if res, err := s.deleteStorageClient(r); err != nil {
+		return reconcile.Result{}, err
+	} else if !res.IsZero() {
+		return res, nil
+	}
+
+	return reconcile.Result{}, nil
+}
+
+func (s *storageClient) createStorageClient(r *StorageClusterReconciler, storagecluster *ocsv1.StorageCluster) (reconcile.Result, error) {
+
+	storageClientCRD := &extv1.CustomResourceDefinition{}
+	storageClientCRD.Name = storageClientCRDName
+	err := r.Get(r.ctx, client.ObjectKeyFromObject(storageClientCRD), storageClientCRD)
+	if err != nil {
+		if kerrors.IsNotFound(err) {
+			r.Log.Info("Waiting for StorageClient CRD to be present")
+			// if we are running in provider mode, deployment of storageclient is mandatory and so poll for crd
+			return reconcile.Result{RequeueAfter: 1 * time.Minute}, nil
+		}
+		r.Log.Error(err, "Failed to verify existence of StorageClient CRD")
+		return reconcile.Result{}, err
+	}
```

Comment:
Doing a hard check on the existence of the CRD means that if the ocs-client-operator (and CRDs) is not installed then the reconciliation of a StorageCluster in Provider mode will never succeed without it. This seems like a bad idea.

Reply:
Yes, that's the intention, w/o local storageclient there'll no usable storageclasses eventually.

Reply:
removed.

Reply:
...that's still bad. We shouldn't be *breaking* the current model, there's no technical reason for this limitation to be introduced. I'd want to see a boolean or something, maybe a new section under `managedResources` would suffice.

Reply:
Sry, I didn't get the ask, we are deploying storageclient if we are running in provider mode.